### PR TITLE
Add DateTimeOffset parameter check and some tests

### LINKED ISSUE #148: Analyzer update needed for string->datetime offset conversion in testcaseattribute
NUnit Version: 3.13

Support was added for string->datetime offset conversion in https://github.com/nunit/nunit/issues/3053

But I've noticed I'm getting an analyzer info message about an invalid conversion when using it. We should update the analyzers to be aware of the new type conversion.

Assuming we want to go ahead with this change, I'd actually be happy to try to implement too. Would be a much-wanted introduction for me to Roslyn analyzers.

### PR DESCRIPTION
**Pull request by stevenaw**
Fixes #148 

This PR adds support for detecting valid string -> DateTimeOffset conversions in TestCase.

### CODE REVIEW

File: src/nunit.analyzers/Extensions/AttributeArgumentSyntaxExtensions.cs [hunk 1]
```diff
@@ -64,6 +64,11 @@ internal static bool CanAssignTo(this AttributeArgumentSyntax @this, ITypeSymbol
                     {
                         canConvert = TimeSpan.TryParse(argumentValue as string, out _);
                     }
+                    else if (argumentType.SpecialType == SpecialType.System_String &&
+                        model.Compilation.GetTypeByMetadataName(typeof(DateTimeOffset).FullName).IsAssignableFrom(targetType))
+                    {
+                        canConvert = DateTimeOffset.TryParse(argumentValue as string, out _);
```
Comment (stevenaw) [inline src/nunit.analyzers/Extensions/AttributeArgumentSyntaxExtensions.cs]: Note to reviewer: I used `TryParse` here, but I think the underlying NUnit implementation uses `TypeConverter.CanConvertFrom`. Please let me know if you'd prefer I update it.

Reference code: https://github.com/nunit/nunit/blob/master/src/NUnitFramework/framework/Internal/ParamAttributeTypeConversions.cs#L138-L143

File: src/nunit.analyzers.tests/TestMethodUsage/TestMethodUsageAnalyzerTests.cs [hunk 2]
```diff
@@ -51,6 +51,29 @@ public void VerifySupportedDiagnostics()
                 $"{TestMethodUsageAnalyzerConstants.SpecifiedExpectedResultForVoidMethodMessage} is missing.");
         }
 
+
+        private static IEnumerable<TestCaseData> SpecialConversions
```
Comment (stevenaw) [inline src/nunit.analyzers.tests/TestMethodUsage/TestMethodUsageAnalyzerTests.cs]: Note to reviewer: These have been duplicated across test files because I couldn't find a good place to centralize them. Please let me know if you have any suggestions, or think its ok as-is.

File: src/nunit.analyzers.tests/TestCaseUsage/TestCaseUsageAnalyzerTests.cs [hunk 3]
```diff
@@ -100,6 +110,30 @@ public sealed class AnalyzeWhenArgumentIsCorrect
...
+    public sealed class AnalyzeWhenArgumentIsSpecialConversion
+    {
+        [TestCase(""" + value + @""")]
+        public void Test(" + targetType.Name + @" a) { }
+    }");
+            AnalyzerAssert.Valid<TestCaseUsageAnalyzer>(testCode);
+        }
+        
+        [Test]
+        public void AnalyzeWhenArgumentIsVerySpecialConversion()
```
Comment (mikkelbu) [inline src/nunit.analyzers.tests/TestCaseUsage/TestCaseUsageAnalyzerTests.cs]: This test looks the same as the `TestCaseData("2019-10-10", typeof(DateTimeOffset))` case above?

File: src/nunit.analyzers.tests/TestCaseUsage/TestCaseUsageAnalyzerTests.cs [hunk 4]
```diff
@@ -15,6 +15,16 @@ namespace NUnit.Analyzers.Tests.TestCaseUsage
...
     {
         private DiagnosticAnalyzer analyzer = new TestCaseUsageAnalyzer();
+        
+        private static IEnumerable<TestCaseData> SpecialConversions
+        {
+            get
+            {
+                yield return new TestCaseData("2019-10-10", typeof(DateTime));
+                yield return new TestCaseData("23:59:59", typeof(TimeSpan));
+                yield return new TestCaseData("2019-10-10", typeof(DateTimeOffset));
```
Comment (mikkelbu) [inline src/nunit.analyzers.tests/TestCaseUsage/TestCaseUsageAnalyzerTests.cs]: Can you also provide a testcase where we have time and offset for the `DateTimeOffset`?

File: src/nunit.analyzers.tests/TestMethodUsage/TestMethodUsageAnalyzerTests.cs (hunk 2 above)
Comment (mikkelbu) [reply src/nunit.analyzers.tests/TestMethodUsage/TestMethodUsageAnalyzerTests.cs]: I think it is fine as it is, but perhaps also add a testcase with a time and offset for `DateTimeOffset`.

File: src/nunit.analyzers/Extensions/AttributeArgumentSyntaxExtensions.cs (hunk 1 above)
Comment (mikkelbu) [reply src/nunit.analyzers/Extensions/AttributeArgumentSyntaxExtensions.cs]: Actually, I don't know the exact differences between `DateTimeOffset.TryParse` and `TypeConverter.CanConvertFrom`, so I'm unsure about this. But I think it will be the safest to have the same implementation, if it is not too cumbersome. Also I think we should use `CultureInfo.InvariantCulture`.

File: src/nunit.analyzers.tests/TestCaseUsage/TestCaseUsageAnalyzerTests.cs (hunk 3 above)
Comment (stevenaw) [reply src/nunit.analyzers.tests/TestCaseUsage/TestCaseUsageAnalyzerTests.cs]: Oh you're right, thanks for spotting this! I'd been rushing to get this PR up before my laptop battery died, and I must've missed cleaning up this earlier attempt at a test.

File: src/nunit.analyzers/Extensions/AttributeArgumentSyntaxExtensions.cs (hunk 1 above)
Comment (stevenaw) [reply src/nunit.analyzers/Extensions/AttributeArgumentSyntaxExtensions.cs]: Thanks for the feedback. I think I have this working now using `TypeConverter` + `CultureInfo.InvariantCulture` :)

**Comment by stevenaw** [comment]
Thanks for the review @mikkelbu 
I think I've covered all the feedback. I've also updated the tests I added to use string interpolation (instead of concatenation) since that seems to be what most of the tests do. I've also added 2 test cases to the existing `AttributeArgumentSyntaxExtensionsTests` to cover the positive/negative `DateTimeOffset`-related tests
